myia/ir.py

```diff
@@ -1,7 +1,29 @@
-"""Intermediate representation definition."""
+"""Base interface for graph-based IR.
 
+This module defines the basic interface for the graph-based IR(s) used in Myia.
 
-class Node:
-    """A node in the graph-based ANF IR."""
+"""
+from abc import ABC, abstractmethod
+from typing import Iterable
 
-    pass
+
+class Node(ABC):
+    """A node in a graph IR.
+
+    Attributes:
+        incoming: A collection of nodes which are inputs to this node.
+        outgoing: A collection of nodes that use this node as an input.
+
+    """
+
+    @property
+    @abstractmethod
+    def incoming(self) -> Iterable['Node']:
+        """Return nodes used as inputs."""
+        pass
+
+    @property
+    @abstractmethod
+    def outgoing(self) -> Iterable['Node']:
+        """Return nodes which use this node as an input."""
+        pass
```

Comment:
I would add `incoming_edges` and `outgoing_edges` to iterate over label/node/maybe_debug_info instead of just nodes, but we can add that if/when it's actually needed.

Reply:
You mean that `incoming_edges` would return `(index, node)` tuples for the core IR and `(label, node)` tuples for the extended one? Let's consider that later, because that violates substitutability again because the return types would be different for different subclasses (mypy wouldn't accept it, for good reasons).

Reply:
The generic type would be `Iterable[Tuple[Any, Node]]`. It is fine for subclasses to return any subtype of that, and mypy does accept this, although that's not saying much, because it seems to accept it the other way around as well (which is incorrect). But yeah let's just discuss that another time.